TFNetworkRecLayer.py

```diff
@@ -4383,3 +4383,273 @@ def __call__(self, inputs, state, scope=None):
+      self._add_params(params=params, scope_name_prefix=scope_name_prefix)
+
+  @classmethod
+  def _post_init_output(cls, output, network, target=None, size_target=None, sources=(), **kwargs):
+    """
+    :param Data output:
+    :param TFNetwork.TFNetwork network:
+    :param str|None target:
+    :param str|None size_target:
+    :param list[LayerBase] sources:
+    """
+    if network.train_flag is not False:
+      # TODO: In training, this is ok. Maybe as well as for eval but not clear.
+      # In forward, mark_data_key_as_used=False should be used and anyway that target value is not available.
+      output.size_placeholder = cls._static_get_target_value(
+        target=target or size_target, network=network,
+        mark_data_key_as_used=network.train_flag is not False).size_placeholder.copy()
+    else:
+      with tf.control_dependencies(None):
+        output.size_placeholder = {} # Not sure why this has to be defined. Is it ever used?
+      pass
+    if any([(not src.output.available_for_inference) for src in sources]):
+      output.available_for_inference = False
+
+  def _add_params(self, scope_name_prefix, params):
+    """
+    :param str scope_name_prefix:
+    :param list[tf.Variable] params:
+    """
+    for p in params:
+      if not p.name.startswith(scope_name_prefix):
+        continue
+      assert p.name.startswith(scope_name_prefix) and p.name.endswith(":0")
+      self.params[p.name[len(scope_name_prefix):-2]] = p
+
+  def _get_input(self):
+    """
+    :return: (x, seq_len), where x is (time,batch,...,dim) and seq_len is (batch,)
+    :rtype: (tf.Tensor, tf.Tensor)
+    """
+    assert len(self.sources) == 2
+    assert self.sources[0].output
+    assert self.sources[1].output
+    x = self.sources[0].output.get_placeholder_as_time_major()  # (time,batch,[dim])
+    seq_len_src = self.sources[0].output.get_sequence_lengths()
+
+    return x, seq_len_src
+
+  def get_constraints_value(self):
+    v = super(TwoDLSTMLayer, self).get_constraints_value()
+    from TFUtil import optional_add
+    if isinstance(self.cell, _SubnetworkRecCell):
+      for layer in self.cell.net.layers.values():
+        v = optional_add(v, layer.get_constraints_value())
+    return v
+
+  def _get_cell(self, unit_opts=None):
+    """
+    :param None|dict[str] unit_opts:
+    :rtype: _SubnetworkRecCell|tensorflow.contrib.rnn.RNNCell|tensorflow.contrib.rnn.FusedRNNCell|TFNativeOp.RecSeqCellOp
+    """
+    import TFNativeOp
+    rnn_cell_class = TFNativeOp.TwoDNativeLstmCell
+    n_hidden = self.output.dim
+    if unit_opts is None:
+      unit_opts = {}
+
+    assert not self.sources[0].output.sparse
+    n_input_dim_parts = [self.sources[0].output.dim, self.sources[1].output.dim]
+    cell = rnn_cell_class(
+      n_hidden=n_hidden, n_input_dim=sum(n_input_dim_parts), n_input_dim_parts=n_input_dim_parts,
+      input_is_sparse=self.sources[0].output.sparse,
+      **unit_opts,
+      pooling=self.pooling)
```

Comment:
I think this is Python 3 syntax, right? The code should also work (or at least load without syntax error) in Python 2 (at least for a while).